sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,530 @@
+`get` to be invoked implicitly, the property `balanced` may be specified on a
+`task_sequence` object, which guarantees that a user will not allow a destructor
+on that `task_sequence` object to be called when there are outstanding `async`
+invocations that have not been balanced by a matching `get` call. On spatial
+architectures, in the presence of this property, potentially expensive hardware
+may be elided. It is undefined behavior to specify the `balanced` property on
+`task_sequence` and then to allow the `task_sequence` object to be destroyed
+while there are any `async` invocations for which `get` has not been called.
+
+== `task_sequence` Properties
+
+The following code and table describe the properties that can be provided when
+declaring a `task_sequence` object.
+
+[source,c++,linenums]
+----
+namespace sycl::ext::intel::experimental {
+struct balanced_key {
+  using value_t = property_value<balanced_key>;
+};
+
+struct invocation_capacity_key {
+  template <uint32_t Size>
+  using value_t = property_value<invocation_capacity_key,
+    std::integral_constant<uint32_t, Size>>;
+};
+
+struct response_capacity_key {
+  template <uint32_t Size>
+  using value_t = property_value<response_capacity_key,
+    std::integral_constant<uint32_t, Size>>;
+};
+
+inline constexpr balanced_key::value_t balanced;
+template <uint32_t Size>
+inline constexpr invocation_capacity_key::value_t<Size> invocation_capacity;
+template <uint32_t Size>
+inline constexpr response_capacity_key::value_t<Size> response_capacity;
+
+template <> struct is_property_key<balanced_key> : std::true_type {};
+template <> struct is_property_key<invocation_capacity_key> : std::true_type {};
+template <> struct is_property_key<response_capacity_key> : std::true_type {};
+
+template <auto &f, typename PropertyListT>
+struct is_property_key_of<balanced_key, task_sequence<f, PropertyListT>>
+  : std::true_type {};
+template <auto &f, typename PropertyListT>
+struct is_property_key_of<invocation_capacity_key, task_sequence<f, PropertyListT>>
+  : std::true_type {};
+template <auto &f, typename PropertyListT>
+struct is_property_key_of<response_capacity_key, task_sequence<f, propertiesT>>
+  : std::true_type {};
+
+} // namespace sycl::ext::intel::experimental
+----
+
+--
+[options="header"]
+|===
+| Property | Description
+| balanced | The `balanced` property is a guarantee to the SYCL device compiler
+that the `task_sequence` object will call exactly the same number of `async` s
+and `get` s over the object's lifetime (i.e. before the `task_sequence`
+destructor is invoked).
+
+| invocation_capacity | The `async` invocations are guaranteed to not block the
+caller as long as the number of outstanding invocations are less than or equal
+to `invocation_capacity`. An outstanding invocation is an inflight
+`task_sequence` invocation that has not yet completed.
+
+A default value is chosen by the implementation.
```

Comment:
This is a compile time property that sets the depth of the generated hardware FIFOs on the FPGA to accommodate extra async'ed invocations and results ready for retrieval, so users would already know if they need to add any of these capacities as properties, or if they don't care about them. A user wouldn't use a query of this value to decide different kinds of hardware to generate unless they already knew the performance impacts that different capacities would have, in which case they would have already specified the property with a compile time argument and used that to decide what upstream/downstream hardware to instantiate instead of relying on the default value of the implementation.

Reply:
Thanks for explanation, I guess I wasn't thinking about having a query for deciding whether to use a particular hardware, it was more to handle the case where a user needs to decide whether the number of `async`/`get` calls they are making will exceed the default limit and would therefore need to specify these properties. I guess this information could be provided by the hardware documentation, but perhaps having a query could allow users to adapt the implementation dynamically.